### PR TITLE
lnurl-auth support

### PR DESCRIPTION
**Pull request by robbiehanson**
This PR builds on previous commits from @dpad85, and adds the UI for lnurl-auth.

lnurl-auth login

### CODE REVIEW

File: phoenix-ios/phoenix-ios/views/SendView.swift [hunk 1]
```diff
@@ -733,7 +737,252 @@ struct SendingView: View {
...
+				}
+				.font(.title2)
+				.foregroundColor(Color.appPositive)
+				.padding(.top, 4)
+				.padding(.bottom, 5)
+				.padding([.leading, .trailing], 24)
+				
+			} else {
+				
+				Button {
```
Comment (dpad85) [inline phoenix-ios/phoenix-ios/views/SendView.swift:828]: This button looks a bit strange, maybe larger than usual (for example compared to the pay button). Or is it just me ? 👀

File: phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/managers/LNUrlManager.kt [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            httpClient.get(url)
+        } catch (sre: io.ktor.client.features.ServerResponseException) {
+            // ktor throws an exception when we get a non-200 response from the server.
+            // That's not what we want. We'd like to handle the JSON error ourselves.
+            sre.response
+        } catch (t: Throwable) {
+            throw LNUrl.Error.RemoteFailure.CouldNotConnect(origin = url.host)
+        }
+
+        LNUrl.handleLNUrlResponse(response) // throws on any/all non-success
```
[the remark this answers is not in the document]
Comment (robbiehanson) [reply phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/managers/LNUrlManager.kt:141]: `LNUrl.Auth.Action` is used by the UI when showing the button. So the button will be a localized version of `Authenticate`, `Log In`, `Register`, or `Link`.

In my first version, I actually had this method return a value of type `LNUrl.Auth.Result`, which consisted of `Authenticated`, `Logged In`, `Registered` & `Linked`. But it seemed a bit unnecessary, as the UI could simply map to these values itself via `LNUrl.Auth.Action`. So I ended up removing it.

File: phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/managers/LNUrlManager.kt [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        // > `login.site.com` will result in different account for each user because
+        // > full domain name is used by wallets as material for key derivation.
+        // >
+        // > LN SERVICE should consider giving meaningful names to chosen subdomains
+        // > since LN WALLET may show a full domain name to users on login attempt.
+        // > For example, `auth.site.com` is less confusing than `ksf03.site.com`.
+        //
+        // Spec: https://github.com/fiatjaf/lnurl-rfc/blob/luds/04.md
+        //
+        val domain = auth.url.host
```
[the remark this answers is not in the document]
Comment (robbiehanson) [reply phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/managers/LNUrlManager.kt:119]: That's a good point. Web services will undoubtably change their sub-domain. And if they do so while lnurl-auth is young, and has relatively few users on their site (compared to other offered login options, such as Facebook), then the website is unlikely to fix their mistake. The wallet will surely receive the blame, both by the web service and the user. And we will gain little ground in the argument by deferring to the spec.

> eTLD+1

This sounds like a logical solution. And, from a user's perspective, is probably what I would expect.

I will work on adding code to fetch & parse the public list.

File: phoenix-ios/phoenix-ios/views/SendView.swift (hunk 1 above)
Comment (robbiehanson) [reply phoenix-ios/phoenix-ios/views/SendView.swift:831]: Done in 163dc42

**Comment by robbiehanson** [comment]
Support for `Public Suffix List` has been added to shared module. Here's how it works:

- When ScanController detects a lnurl-auth, it starts a pre-fetch of the Public Suffix List (PSL)
- When asked to authenticate, the PSL is consulted in order to extract the eTLD+1
- The fetched PSL is cached in the database (and only re-fetched if older than 24 hours)

Unit tests have been added for the PSL logic. Some notes:

The PSL list is really big. It would probably be more performant if we cached this as a file. However, I think we need a 3rd party library to do basic filesystem IO in Kotlin Native. (Maybe [Okio](https://square.github.io/okio/multiplatform/)) So the simpler solution was to store it in the database. It would be nice to compress the data we store in the database. But I think we need a 3rd party library to do that too...

Since we're adding another table to the database, I took this opportunity to add a simple key/value table that can be re-used for other things in the future.

I had to disable some unit tests because we aren't performing proper IDN on the domain name. In particular, there is a set of unit tests that require punycode support. It looks like this exists in `java.net.IDN`, but not in Kotlin native.

File: phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/managers/LNUrlManager.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (robbiehanson) [reply phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/managers/LNUrlManager.kt]: Done in c035bf0

File: phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/utils/PublicSuffixList.kt [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+                }
+
+                var isExceptionRule = false
+                if (suffix.startsWith('!')) {
+                    isExceptionRule = true
+                    suffix = suffix.substring(1)
+                }
+
+                val labels = suffix.split('.').filter {
+                    it.isNotEmpty()
```
[the remark this answers is not in the document]
Comment (robbiehanson) [reply phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/utils/PublicSuffixList.kt]: Done in 83b5d3f

File: phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/utils/PublicSuffixList.kt [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+                // - Each line which is not entirely whitespace or begins with a comment contains a rule.
+                // - A rule may begin with a "!" (exclamation mark).
+                //   If it does, it is labelled as an "exception rule"
+                //   and then treated as if the exclamation mark is not present.
+                // - A domain or rule can be split into a list of labels using the separator "." (dot).
+                //   The separator is not part of any label.
+                //   Empty labels are not permitted, meaning that leading and trailing dots are ignored.
+
+                var suffix = line.split(regex = whitespace, limit = 1).firstOrNull() ?: ""
+                if (suffix.isEmpty() || suffix.startsWith("//")) {
```
[the remark this answers is not in the document]
Comment (robbiehanson) [reply phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/utils/PublicSuffixList.kt]: Done in 83b5d3f

File: phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/managers/AppConfigurationManager.kt [hunk 6]
```diff
@@ -107,7 +131,44 @@ class AppConfigurationManager(
         }
     }
 
-    /** The flow containing the configuration to use for Electrum. If null, we do not know what conf to use. */
+    private val publicSuffixListKey = "publicSuffixList"
+
+    public suspend fun fetchPublicSuffixList(
+        refreshIfOlderThan: Duration = 1.days
```
[the remark this answers is not in the document]
Comment (robbiehanson) [reply phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/managers/AppConfigurationManager.kt]: Good catch. Done in 83b5d3f

**Review by dpad85 — APPROVED** [review]
LGTM 👍

**Pull request merged**